File: qualification/elixir/lib/calls.ex
defmodule LspTraceQualification.Calls do
  # Multi-clause target: qualification must resolve callers to one logical leaf/1 target.
  def leaf(value) when is_integer(value), do: value + 1
  def leaf(value), do: value

  def left(value), do: leaf(value)
  def right(value), do: leaf(value)
  def same_module(value), do: leaf(value)

  def recursive(value) when value <= 0, do: leaf(value)
  def recursive(value), do: leaf(value) + recursive(value - 1)

  # Static-only witness: the normal entry path never takes this branch.
  def static_but_not_executed(value) do
    if System.get_env("LSP_TRACE_NEVER_SET") == "execute", do: leaf(value), else: value
  end

  def entry(value), do: left(value) + right(value) + recursive(value)
end

File: qualification/csharp/Calls.cs
namespace LspTraceQualification;

public static class Calls
{
    public static int Leaf(int value) => value + 1;
    public static int Left(int value) => Leaf(value);
    public static int Right(int value) => Leaf(value);
    public static int Recursive(int value) => value <= 0 ? Leaf(value) : Recursive(value - 1);

    // Static-only witness: normal fixture execution never takes this branch.
    public static int StaticButNotExecuted(int value) =>
        Environment.GetEnvironmentVariable("LSP_TRACE_NEVER_SET") == "execute" ? Leaf(value) : value;

    public static int Entry(int value) => Left(value) + Right(value) + Recursive(value);
    public static void Main() => Console.WriteLine(Entry(2));
}
